[Extensions] Add DynamicActionRegistry to ActionModule

dbwiddis opened the pull request:
Companion PR: [SDK #588](https://github.com/opensearch-project/opensearch-sdk-java/pull/588).  This PR must be merged first.

### Description

Adds a `DynamicActionRegistry` class and instance field to the `ActionModule`.

TransportAction classes from OpenSearch and plugins are instantiated during Node bootstrap and injected as singletons.  

Extensions have the need to proxy actions between extensions. For example, job scheduler needs to invoke actions on Anomaly Detector extension; Anomaly Detector may need to invoke actions on Alerting Extension; et. al.   This capability is currently implemented with a single `ProxyAction` which relays these invocations blindly.

This PR adds a registry for Extensions to register and unregister classes which extend `ActionType` and `TransportAction` at times other than bootstrap.

The benefit of using a subclass of `TransportAction` is that the `client.execute()` invocation processes these actions through ActionFilters.  The first such filter processed is in the security plugin and this provides a single point of contact for future application of security features.

Inspiration for some of these features: #4460 (but I did it [my way](https://www.youtube.com/watch?v=w019MzRosmk)) 

Description of functionality:
 - On bootstrap, `ActionModule` creates an registry with two internal maps; one corresponds to the `actions` map obtained from the injector, which is immutable, and an empty registry for addition (and potentially removal) of extension actions
 - During Extension initialization, the `getActions()` extension point will be handled on OpenSearch by the `ExtensionTransportActionsHandler` which will register (in the registry) a custom subclass of the `ActionType` containing the fully qualified class name of the action on the Extension
   - There is no current usage of the implemented "unregister" functionality, but the capability is provided for future "hot swap" functionality.
 - When one extension (or OpenSearch or a plugin) wants to invoke an action on an extension, that will be handled by the NodeClient's `client.execute()` path. This path uses associated ActionFilters and the potential for future checks and security/identity integration.
 - The execution of this `TransportAction` will send it to the Extension for further processing (TBD in Companion PR).

### Issues Resolved

Fixes [SDK #107](https://github.com/opensearch-project/opensearch-sdk-java/issues/107).

Companion PR implements [SDK #525](https://github.com/opensearch-project/opensearch-sdk-java/issues/525).

[SDK #525](https://github.com/opensearch-project/opensearch-sdk-java/issues/525)  could be implemented using the existing `ProxyAction` setup that this PR replaces, but I believe the benefits of funneling all requests through a single touch point is a significant benefit and combined these efforts.

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

peternied commented on server/src/main/java/org/opensearch/client/node/NodeClient.java:
```diff
@@ -82,6 +93,21 @@ public void initialize(
+        NamedWriteableRegistry namedWriteableRegistry
+    ) {
+        initialize(actions, localNodeId, transportService.getRemoteClusterService(), namedWriteableRegistry);
+        this.extensionActions = actionModule.getExtensionActions();
```
I don't like the abstraction of 'action' vs 'extensionAction' leaking into the NodeClient.  What do you think about encapsulating the extensionAction concepts fully inside of ActionModule?

dbwiddis replied:
So I was able to refactor and get everything into a single registry.  Haven't tested whether it works but we'll see! :)

peternied replied:
🤞

peternied commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -722,6 +731,13 @@ public <Request extends ActionRequest, Response extends ActionResponse> void reg
         return unmodifiableMap(actions.getRegistry());
     }
 
+    public DynamicActionRegistry getExtensionActions() {
```
Why should extension actions be treated differently by callers of ActionModule? Seems like they are all still actions

dbwiddis replied:
I've combined the two maps (immutable and dynamic) inside the registry so they're an implementation detail. I think that tries to address this where external to the module they can be treated the same.

peternied reviewed (commented):
Nice! I like the steps forward to consolidating all actions together.  I've put in some notes about coupling and how the abstraction could work a little differently that I think could be more flexible.

cwperks commented on server/src/main/java/org/opensearch/extensions/action/ExtensionTransportActionsHandler.java:
```diff
@@ -64,10 +69,12 @@ public ExtensionTransportActionsHandler(
         }
-        actionsMap.putIfAbsent(action, extension);
+        // Register the action in the action module's extension actions map
+        dynamicActionRegistry.registerExtensionAction(new ExtensionAction(action, extension.getId()));
```
If the action requires parameters then how will then be sent to the extension?

cwperks replied:
+1 for protobufs

peternied commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -954,13 +965,104 @@ protected void configure() {
+     *
+     * @opensearch.internal
+     */
+    public static class DynamicActionRegistry<Request extends ActionRequest, Response extends ActionResponse> {
```
Do these generics do anything? Looks like in its usage such as NodeClient they aren't provided

dbwiddis replied:
I have been fighting with those generics for at least an hour.  Without them, the `get()` call in NodeClient fails.

If you know how to properly do them or can point me to a "Generics 101" tutorial, I'd buy you a green beverage of choice.

dbwiddis replied:
Things work when I keep the maps separate but trying to have a combined get method that includes subclasses is not.

peternied reviewed (commented):
Thanks for the refactor 👍

owaiskazi19 commented on server/src/main/java/org/opensearch/client/node/NodeClient.java:
```diff
@@ -137,10 +143,11 @@ public String getLocalNodeId() {
         }
-        TransportAction<Request, Response> transportAction = actions.get(action);
+        // Get from action map if it exists
+        TransportAction<Request, Response> transportAction = actionRegistry.get(action);
```
Since we are using feature flag to register actions dynamically. We should not this change directly on `main` branch. 
We can have it like:
```java
 TransportAction<Request, Response> transportAction = actions.get(action);
 if(feature_flag_enabled){
   TransportAction<Request, Response> dynamicTransportAction = actionRegistry.get(action);
 }
 ```

owaiskazi19 replied:
Ah I missed the small `get` method in `ActionModule`. Looks fine. Ignore my above comment

owaiskazi19 commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -471,7 +476,10 @@ public class ActionModule extends AbstractModule {
     private final List<ActionPlugin> actionPlugins;
+    // An unmodifiable map containing OpenSearch and Plugin actions
     private final Map<String, ActionHandler<?, ?>> actions;
+    // A dynamic map combining the above immutable actions with dynamic Extension actions
```
Moreover a design question. Do we plan to remove the immutable map completely for `actions` in the future and use the mutable one?

dbwiddis replied:
No, I don't think so.  See my reply to @peternied [here](https://github.com/opensearch-project/OpenSearch/pull/6734#discussion_r1140418316)

owaiskazi19 replied:
> I could put everything in a single registry, but this would:

  >  1.  remove the existing @Inject dependency action capability and require updating all the code to fetch from the registry instead
  
   Definitely. As we are moving towards dynamic actions, existing @Inject can be removed.
   >  2. make the currently immutable actions map mutable, removing the ability for developers to rely on actions always being available
   
   OpenSearch actions would be still available because they will be registered during boostrap.
  > 3. probably break all existing plugins that rely on this functionality
  
  Since we are moving towards extensions. Features like this would come more often. We can start a campaign for plugins to migrate to the new change.
  
  All I wanted to say let's not say `No` to deprecating old way of registering actions. Dynamic actions is the way to move forward in the future with plugins migrating to extensions and can have only one map to register actions.

dbwiddis replied:
> OpenSearch actions would be still available because they will be registered during boostrap.

But they would be allowed to be unregistered.  That's the problem.

We need to treat "same JVM" vs. "remote JVM" actions differently anyway.  The current split does that well.

owaiskazi19 replied:
Makes sense. Can we add this detail for having 2 maps in the comment for providing more information?

dbwiddis replied:
Sure, I'll expand the above comments.

owaiskazi19 commented:
Overall looks good @dbwiddis. Small ask: would love to see a design diagram for the workflow in the SDK repository. Since we are touching the ActionModule here would be great to get some eyes from @reta @nknize

dbwiddis commented:
Gradle check failures:
 - #5958 
 - #6287 
 - #6738

dbwiddis commented:
> Overall looks good @dbwiddis. Small ask: would love to see a design diagram for the workflow in the SDK repository. Since we are touching the ActionModule here would be great to get some eyes from @reta @nknize

A sequence diagram has been added to the companion PR.  See [here](https://github.com/opensearch-project/opensearch-sdk-java/blob/bee5f3e9a7e9d905775aaff0f5dc41fc8ad64871/Docs/RemoteActionExecution.svg).

joshpalis commented on server/src/main/java/org/opensearch/extensions/action/ExtensionActionResponse.java:
```diff
@@ -52,8 +68,27 @@ public void setResponseBytes(byte[] responseBytes) {
+     *
+     * @return A string representation of the response bytes
+     */
+    public String getResponseBytesAsString() {
```
Perhaps we could also provide a method to wrap the `responseBytes `with a `StreamInput` to allow reading in objects using a stream input constructor.

dbwiddis replied:
Great suggestion! Will do.

saratvemulapalli commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -471,7 +475,17 @@ public class ActionModule extends AbstractModule {
+    // and also registers dynamic ExtensionActions / ExtensionTransportActions
+    // associated with remote action execution on extensions, usually in
+    // a different JVM and possibly on a different server.
+    private final DynamicActionRegistry dynamicActionRegistry;
```
I like how you've separated out dynamic actions differently, what concerns do you see of modifying the existing actions ? I see we aren't using them anyway. 
I can think of inconsistencies with other nodes which might not have registered actions after bootstrap. 

I understand the safeguards make sense for 2.x but for 3.x Im leaning towards having one set of actions which are injected and read by the client.

dbwiddis replied:
@saratvemulapalli "injected" only happens once, and the whole ActionType instance -> TransportAction class setup is tightly integrated.   The injector already has the necessary instances.   Removing them from the map wouldn't take away the singleton that is still bound to that class.    In order to truly "unregister" the things loaded in during bootstrap we need to get away from injection.

dbwiddis replied:
> I can think of inconsistencies with other nodes which might not have registered actions after bootstrap.

This is actually a bigger issue that I hadn't designed for.  What is the easiest existing method for replicating that?   I designed this registry after named registry, does that replicate across nodes?

dbwiddis replied:
OK..... some follow-up comments:
1. Conceptually this just boils down to an immutable map and a mutable one, plus some methods to access it
2. I will do renaming to make it more generic describing the above
3. There is some point that `client.execute()` on the NodeClient needs access to both maps.   We have the choice of adding a separate map and passing both to NodeClient (my original PR before refactoring in response to @peternied comments) or combining them here inside this module.   Or combining them somewhere else in a whole new module.   I think here is the simplest approach, particularly since they are all "actions".

saratvemulapalli replied:
Makes sense thanks @dbwiddis !

saratvemulapalli commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -954,13 +964,107 @@ protected void configure() {
+         *
+         * @param extensionAction The action to add
+         */
+        public void registerExtensionAction(ExtensionAction extensionAction) {
+            requireNonNull(extensionAction, "extension action is required");
+            String name = extensionAction.name();
+            String uniqueId = extensionAction.uniqueId();
+            if (registry.containsKey(extensionAction)) {
+                throw new IllegalArgumentException("extension [" + uniqueId + "] action [" + name + "] already registered");
+            }
+            registry.put(extensionAction, new ExtensionTransportAction(name, actionFilters, transportService, extensionsManager));
+        }
```
hm... I have couple of concerns:
- ActionModule/DynamicActionRegistry shouldn't understand details of extension business logic. 
Do you think putting in an interface of registerActionFilter and let ExtensionManager implement its own filter which would cleanly implement its business logic. 

- This really doesnt support registering actions dynamically for rest of the system. 
Developers would love a way to register these actions dynamically. I can think of plugins leveraging this for hotswap, register actions based on settings toggled etc.

dbwiddis replied:
This is just a naming/typing thing that's easily fixed.  `registerDynamicAction(ActionType action)` would do the same thing, and we could generalize the uniqueID to just some other id field.

saratvemulapalli reviewed (commented):
This is great @dbwiddis, helping OpenSearch to be dynamic one module at a time. 

@reta @nknize would love your thoughts

reta commented:
> @reta @nknize would love your thoughts

My apologies @saratvemulapalli , haven't had a time to look at it yet, but will do shortly, thanks!

dbwiddis commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -954,13 +964,107 @@ protected void configure() {
+         */
+        @SuppressWarnings("unchecked")
+        public TransportAction<? extends ActionRequest, ? extends ActionResponse> get(ActionType<?> action) {
+            if (action instanceof ExtensionAction) {
```
Yeah, I was already thinking with that other comment I should look first in the keyset of the immutable map and then use this map as a fallback. Will address in the same fix.

dbwiddis commented on server/src/main/java/org/opensearch/extensions/action/ExtensionTransportAction.java:
```diff
@@ -10,37 +10,27 @@
  * @opensearch.internal
  */
-public class ExtensionTransportAction extends HandledTransportAction<ExtensionActionRequest, ExtensionActionResponse> {
+public class ExtensionTransportAction extends TransportAction<ExtensionActionRequest, ExtensionActionResponse> {
```
We are only registering the dynamic actions in the node connected to extensions.  The HandledTransportAction, registers for any node, and doesn't provide unregistration features.

So basically we probably eventually need something like this when we move beyond single node connection to extension but for now it provides no benefit and adds complications.

dbwiddis commented on server/src/test/java/org/opensearch/action/ActionModuleTests.java:
```diff
@@ -262,4 +272,77 @@ public List<Route> routes() {
+        dynamicActionRegistry.initialize(testMap, new ActionFilters(Collections.emptySet()), mockTransportService, null);
+
+        // Should contain the immutable map entry
+        assertTrue(dynamicActionRegistry.get(TestAction.INSTANCE) instanceof TestTransportAction);
```
Yes we can. :)

dbwiddis commented on server/src/test/java/org/opensearch/action/ActionModuleTests.java:
```diff
@@ -262,4 +272,77 @@ public List<Route> routes() {
+            fail("Should not have thrown exception registering action: " + e);
+        }
+        // Should have a value
+        assertTrue(dynamicActionRegistry.get(testExtensionAction) instanceof ExtensionTransportAction);
```
Same answer as above.

reta commented on server/src/main/java/org/opensearch/client/node/NodeClient.java:
```diff
@@ -57,26 +60,29 @@
 
-    private Map<ActionType, TransportAction> actions;
+    private DynamicActionRegistry actionRegistry;
+    private ActionFilters actionFilters;
```
`actionFilters` is not used, please remove (along with `transportService` and `extensionsManager`)

dbwiddis replied:
You'd think the yellow-squigglies in my IDE would have reminded me to do that...

reta commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -954,13 +960,87 @@ protected void configure() {
+        // This is the immutable actions map created during node bootstrap, which
+        // will continue to link ActionType and TransportAction pairs from core and plugin
+        // action handler registration.
+        private Map<ActionType, TransportAction> actions = Collections.emptyMap();
```
Please do not use raw type for type with generic type arguments:

```suggestion
        private Map<ActionType<?>, TransportAction<?, ?>> actions = Collections.emptyMap();
```

Same applies for `registry`

dbwiddis replied:
I can change for `registry` but the raw types are required for the `actions` map.   

The existing code sends to NodeClient with this line:
https://github.com/opensearch-project/OpenSearch/blob/f40fa82e3667825fd0ba8d4bb17b6959ff7cc4f1/server/src/main/java/org/opensearch/node/Node.java#L1115

and the NodeClient version of the map (which is now being wrapped inside the registry) is:
https://github.com/opensearch-project/OpenSearch/blob/f40fa82e3667825fd0ba8d4bb17b6959ff7cc4f1/server/src/main/java/org/opensearch/client/node/NodeClient.java#L60

dbwiddis replied:
Or I guess I can add the `<?>` to the existing code....

dbwiddis replied:
Seems to not work:
```
»  org.opensearch.common.inject.ConfigurationException: Guice configuration errors:
»  
»  1) No implementation for java.util.Map<org.opensearch.action.ActionType<?>, org.opensearch.action.support.TransportAction<?, ?>> was bound.
»    while locating java.util.Map<org.opensearch.action.ActionType<?>, org.opensearch.action.support.TransportAction<?, ?>>
```
I think with type erasure we have to use raw types with the injection.

reta replied:
:o this is very surprising, will take a look shorlty

reta commented on server/src/main/java/org/opensearch/extensions/action/ExtensionActionResponse.java:
```diff
@@ -41,7 +48,16 @@ public ExtensionActionResponse(byte[] responseBytes) {
      */
     public ExtensionActionResponse(StreamInput in) throws IOException {
-        responseBytes = in.readByteArray();
+        this.success = in.readBoolean();
```
So the ExtensionActionResponse is already released (in 2.6 at least): https://github.com/opensearch-project/OpenSearch/blob/2.6/server/src/main/java/org/opensearch/extensions/action/ExtensionActionResponse.java


```suggestion
if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
        this.success = in.readBoolean();
}
```
```

dbwiddis replied:
Yes but it's marked `@opensearch.internal` and is only used by Extensions (not yet released) and will be needed for Extensions GA sometime in 2.X.

reta replied:
That's fine for `@opensearch.internal` but bwc tests will fail in mixed clusters, I suspect we don't have any tests for extensions yet

reta commented on server/src/main/java/org/opensearch/extensions/action/ExtensionActionResponse.java:
```diff
@@ -52,8 +68,36 @@ public void setResponseBytes(byte[] responseBytes) {
+
     @Override
     public void writeTo(StreamOutput out) throws IOException {
+        out.writeBoolean(success);
```
```suggestion
if (out.getVersion().onOrAfter(Version.V_3_0_0)) {
        out.writeBoolean(success);
}
```

reta commented on server/src/main/java/org/opensearch/extensions/action/ExtensionTransportAction.java:
```diff
@@ -10,37 +10,27 @@
+        String actionName,
         ActionFilters actionFilters,
-        ClusterService clusterService,
+        TransportService transportService,
```
It seems like we just need a `TaskManager`, not the `TransportService` here:

```suggestion
        TaskManager taskManager,
```

reta commented on server/src/main/java/org/opensearch/extensions/action/ExtensionTransportActionsHandler.java:
```diff
@@ -40,34 +40,48 @@
     private final NodeClient client;
+    private ActionModule actionModule;
+    private final DynamicActionRegistry dynamicActionRegistry;
+    private ExtensionsManager extensionsManager;
```
```suggestion
    private final ExtensionsManager extensionsManager;
```

dbwiddis commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -954,13 +960,87 @@ protected void configure() {
+         *
+         * @param actions The injected map of {@link ActionType} to {@link TransportAction}
+         */
+        public void initialize(Map<ActionType, TransportAction> actions) {
```
Fair enough.  I used the same method name for sending the action map to `NodeClient` but I agree this makes more sense for this registry.

reta replied:
Or may be, just a suggestion, `registerPreconfiguredActionMap`,

dbwiddis commented:
@reta have addressed all the comments except the version check on `ExtensionActionResponse`.  That was previously only used with `ProxyActionRequest` which was deleted.  However, I'm seeing a need to put that back for use by plugins, so I'll be reworking that with a wrapper class for this PR.   Briefly toggling this into draft mode.

dbwiddis commented:
After taking far too long to debug some stupid mistakes, this is ready for review again.  I'll be updating the sequence diagram on the SDK PR for the name changes.  Status:
 - addressed all comments by @reta except for the need to use raw types for the injected action map, I don't see a way around this
 - left `ExtensionActionResponse` as it currently is (so no versioning check needed) and created a new `RemoteExtensionActionResponse` that manages the success boolean.
 - put back in `ProxyAction` using the original `ExtensionActionResponse`.  It's reasonable for extensions to handle the different request separately (e.g., handle an internal OpenSearch or plugin request differently than one coming from some other extension).

Expect one more commit shortly as I fixup some comments and remove some debug logging, but should be GTG. @saratvemulapalli @owaiskazi19 @joshpalis

dbwiddis commented:
Gradle check failures are https://github.com/opensearch-project/OpenSearch/issues/6565

Precommit failure on windows seems unrelated to this PR (':plugins:discovery-ec2:thirdPartyAudit')

dreamer-89 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** UNSTABLE ❕
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.indices.replication.SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh
> ```
> 
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/12769/
> * **CommitID:** [2aad4f2](https://github.com/opensearch-project/OpenSearch/commit/2aad4f23246f2106cdd8d8400b23df625b392812)
>   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Primary mode test assertion failure, tracked in https://github.com/opensearch-project/OpenSearch/issues/6778
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.replication.SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh" -Dtests.seed=C2F45553749C1906 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-SV -Dtests.timezone=Asia/Urumqi -Druntime.java=19

org.opensearch.indices.replication.SegmentReplicationRelocationIT > testRelocateWhileContinuouslyIndexingAndWaitingForRefresh FAILED
    com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=439, name=opensearch[node_t0][generic][T#3], state=RUNNABLE, group=TGRP-SegmentReplicationRelocationIT]
        at __randomizedtesting.SeedInfo.seed([C2F45553749C1906:4E808D644A8AB9FE]:0)

        Caused by:
        java.lang.AssertionError: shard [test-idx-1][0], node[HtrzAXEsSDO4G3yVaz24DQ], relocating [r14izykMTjSinZumB-UzFA], [P], s[RELOCATING], a[id=mjhIXkBpSf2heaofp6r2Vg, rId=uw6PkvrtT6627c7B_lww3w], expected_shard_size[230] is not a primary shard in primary mode
            at __randomizedtesting.SeedInfo.seed([C2F45553749C1906]:0)
            at org.opensearch.index.shard.IndexShard.assertPrimaryMode(IndexShard.java:2341)
            at org.opensearch.index.shard.IndexShard.updateVisibleCheckpointForShard(IndexShard.java:2728)
            at org.opensearch.indices.replication.OngoingSegmentReplications.startSegmentCopy(OngoingSegmentReplications.java:127)
            at org.opensearch.indices.replication.SegmentReplicationSourceService$GetSegmentFilesRequestHandler.messageReceived(SegmentReplicationSourceService.java:157)
            at org.opensearch.indices.replication.SegmentReplicationSourceService$GetSegmentFilesRequestHandler.messageReceived(SegmentReplicationSourceService.java:154)
            at org.opensearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:106)
            at org.opensearch.transport.InboundHandler$RequestHandler.doRun(InboundHandler.java:453)
            at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:806)
            at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
            at java.****/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
            at java.****/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
            at java.****/java.lang.Thread.run(Thread.java:1589)
```

dreamer-89 commented:
> Gradle check failures are #6565
> 
> Precommit failure on windows seems unrelated to this PR (':plugins:discovery-ec2:thirdPartyAudit')

@dbwiddis : Muted this test in main, please rebase against latest changes.

saratvemulapalli commented on server/src/main/java/org/opensearch/action/ActionModule.java:
```diff
@@ -954,13 +967,86 @@ protected void configure() {
+         *
+         * @param actions The injected map of {@link ActionType} to {@link TransportAction}
+         */
+        public void registerUnmodifiableActionMap(Map<ActionType, TransportAction> actions) {
```
Trying to understand, we could've directly passed in the map L516 when `DynamicActionRegistry` is created. Did you want to keep the structure of the injected information ?

dbwiddis replied:
> Trying to understand, we could've directly passed in the map L516 when `DynamicActionRegistry` is created. 

Yep but it's a different map. 

Given `getActions()` this was the simplest approach and the first thing I tried.

The bits we need are the ActionType key and TransportAction singleton instance value.

The signature of `actions` is `Map<String, ActionHandler<?, ?>>` where the ActionHandler does have the ActionType key, but only the _class_ of the TransportAction.  And the only way to fetch that singleton instance that was injected is to pass around the injector (or a map retrieved from the injector, as was done to send the "map we need" to NodeClient and is now inside the DynamicActionRegistry).

> Did you want to keep the structure of the injected information ?

My original implementation kept this map injected as-is and handled the two different maps in the NodeClient.  However, based on initial review it was requested I keep those details out of the client.   

But we still need to provide an ActionType instance key (easy) and fetch a singleton TransportAction value that we can run the `doExecute()` method on.  

We *could* use the class and get the right instance if we keep a copy of the injector inside the registry, and in theory that might help out with the concern over rawtypes.

But that's also new code/patterns and I really tried to keep this as close to the way it already was as I could.

saratvemulapalli commented on server/src/main/java/org/opensearch/extensions/action/ExtensionAction.java:
```diff
@@ -0,0 +1,60 @@
+        return this.uniqueId;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = super.hashCode();
+        result = prime * result + Objects.hash(uniqueId);
+        return result;
+    }
```
This looks interesting, curious who uses `hashCode`

dbwiddis replied:
These are the key to the map with the TransportAction value.  Having matching hashcodes means I can (and do in this code) re-create the instance to be used as a key using just the strings.

saratvemulapalli commented on server/src/main/java/org/opensearch/extensions/action/ExtensionTransportAction.java:
```diff
@@ -10,44 +10,34 @@
-        Settings settings,
-        TransportService transportService,
+        String actionName,
         ActionFilters actionFilters,
-        ClusterService clusterService,
+        TaskManager taskManager,
```
I am missing where actionFilters and taskManager are being used.

dbwiddis replied:
The `TransportAction` superclass: `super(actionName, actionFilters, taskManager);`

The taskManager maintains the thread context and the action filters are chained during execution of the action.

saratvemulapalli reviewed (approved):
Overall the changes look good to me.
